packages/runtime/container-runtime/src/runningSummarizer.ts

```diff
@@ -329,18 +350,77 @@ export class RunningSummarizer implements IDisposable {
             // Another summary is currently being generated.
             return { alreadyRunning: this.generator.waitSummarizing() };
         }
-        const result = this.generator.summarize(`onDemand;${reason}` as `onDemand;${string}`, options);
-        result.receivedSummaryAckOrNack.finally(() => this.checkRerunHeuristics());
+        const result = this.generator.summarize(onDemandReason, options);
+        result.receivedSummaryAckOrNack.finally(() => this.checkSummarizeAgain());
         return result;
     }
 
-    /** After summarizing, we should rerun the heuristics to see if we already need to summarize again. */
-    private checkRerunHeuristics() {
-        if (this.tryWhileSummarizing) {
+    public enqueueSummarize({
+        reason,
+        afterSequenceNumber = 0,
+        ...options
+    }: IEnqueueSummarizeOptions): ISummarizeResults {
+        const onDemandReason = `enqueue;${reason}` as const;
+        // If afterSequenceNumber is provided, the on-demand summary is "enqueued" to run
+        // once an eligible op comes in with sequenceNumber >= afterSequenceNumber.
+        if (this.enqueuedSummary !== undefined) {
+            this.enqueuedSummary.reason = onDemandReason;
+            this.enqueuedSummary.afterSequenceNumber = afterSequenceNumber;
+            this.enqueuedSummary.options = options;
+        } else {
+            this.enqueuedSummary = {
+                reason: onDemandReason,
+                afterSequenceNumber,
+                options,
+                resultsBuilder: new SummarizeResultBuilder(),
+            };
+        }
+        const results = this.enqueuedSummary.resultsBuilder.build();
+        this.tryRunEnqueuedSummary();
+        return results;
+    }
+
+    /**
+     * After summarizing, we should check to see if we need to summarize again.
+     * Rerun the heuristics and check for enqueued summaries.
+     */
+    private checkSummarizeAgain() {
+        if (this.tryRunEnqueuedSummary()) {
+            this.tryWhileSummarizing = false;
+        } else if (this.tryWhileSummarizing) {
```

Comment:
We should run the `if (this.tryWhileSummarizing)` if we did not run an enqueued summary right?
Consider this scenario - We are running `summarize` and another `summarize` was queued because of heuristics. `checkSummarizeAgain` runs but it doesn't summarize (for whatever reason). We should run another `summarize` at this point.
Also, even if an `enqueuedSummary` runs in the above scenario, don't we want to run the one because of the heuristics as the parameter might be different?

Reply:
Oh, `tryRunEnqueuedSummary` returns false if it didn't summarize. I missed that :-). So, the first part of my question is not valid.

Reply:
I don't think the differing parameters matter right now (2nd part of your question).

Reply:
Yeah, that makes sense. The only interesting one if `fullTree` and that is just an optimization anyway.